pandas/tests/indexes/datetimes/test_arithmetic.py

```diff
@@ -41,6 +41,25 @@ def addend(request):
     return request.param
 
 
+class TestDatetimeIndexComparisons(object):
+    @pytest.mark.parametrize('other', [None,
+                                       datetime(2016, 1, 1).date()])
+    def test_dti_cmp_invalid(self, tz, other):
+        # GH#19301
+        dti = pd.date_range('2016-01-01', periods=2, tz=tz)
```

Comment:
So is this saying that

```python
DatetimeIndex([None, '2016-01-01']) == [None, datetime.date(2016, 1, 1)])
```
is `[False, False`]? I thought that in https://github.com/pandas-dev/pandas/pull/18188 we decided that `DatetimeIndex` compared with `datetime.date` would coerce the `date` to a `datetime` at midnight?

Reply:
@TomAugspurger I think part of the confusion is over `Timestamp` comparison vs `DatetimeIndex` comparison vs `Series[datetime64]` comparison (e.g. the whatsnew note in #18188 talks about `Series` (but puts the tests in index tests...)).  This PR and a bunch of other recent ones have been focused on making the `Index`/`Series` behavior more consistent.

Following this, #19288 can be de-kludged to make `Series[datetime64]` comparison dispatch to `DatetimeIndex` comparison, ensuring internal consistency.  But you're right that this would mean a change in the behavior of `Series[datetime64]` comparisons.

For the moment I'm taking `Timestamp` behavior as canonical and making `DatetimeIndex` match that.

```
ts = pd.Timestamp('2016-01-01')
>>> ts == ts.to_pydatetime().date()
False
>>> ts < ts.to_pydatetime().date()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pandas/_libs/tslib.pyx", line 1165, in pandas._libs.tslib._Timestamp.__richcmp__
TypeError: Cannot compare type 'Timestamp' with type 'date'
```

I recall a discussion of whether `date` should be treated as datetime-at-midnight for `Timestamp` comparison purposes, my thought being it should be treated as `Period(..., freq='D')`.

Reply:
> I find [5] a bit odd.

The behavior is analogous to comparing `Timestamp` vs `Period`.  eq and ne return False and True, respectively, and all others raise `TypeError`.  Its odd if you interpret date as "datetime implicitly at midnight", but pretty intuitive if you interpret it as "less specific than a timestamp"